# HW 03 - Part 2: Data Exploration
DATA 602, Section PCS2, Dr. Fardina Alam


---
---
**DISCLAIMER: This is a fake story, not a real incident! We have created this story just for the data exploration purpose.**

---
---

**Campus Police Department**

Incident Report

---

**Incident Date:** September 10, 2024

**Location of Incident:** 2nd Floor, Iribe Building

**Items Left at the Scene:** Red Ticket for Zichao's Magic Show

**Case Number:** 2024-IRB-0027

---
**Victim Information:**

**Name:** Dr. Fardina Alam

**Affiliation:** University of Maryland, College Park

**Position:** Professor

**Contact Information:** fardina@umd.edu

---

**Incident Description:**

On the evening of September 10, 2024, an unknown individual(s) unlawfully entered Dr. Fardina's room located on the 3rd floor of Iribe and stole a set of Markers. The stolen items were a set of limited edition markers that Dr. Fardina had spent years collecting. At the scene of the crime, a red entry ticket for "Zichao's Magic Show" was found on the floor near Dr. Fardina's desk.







In [1]:
# We import all essential libraries for this HW; if you want to add additional libraries, you can add more here!
import pandas as pd
import numpy as np
from datetime import datetime

## Instructions:

In this assignment, you are tasked with solving the mystery of Dr. Fardina's stolen markers and finding the culprit.

* To accomplish this, you will be provided with **4 data sets** containing key information related to the incident. The list of the data sets name given below:

 - Profiles
 - Witnesses
 - Zichao's magic show
 - Interrogation Statements



* Your goal is to analyze these data sets to uncover the clues and evidences that will lead you to the thief.

**Be careful in following instructions as it may lead to you losing points :(**

Take a moment to look at all the data sets below:

---

**Profiles:**

This dataset contains information regarding all the potential people that could have committed the crime. General information regarding each person are as follows:
- **ID:** A **unique** identification for each person
- **Name:** The name of the person
- **Birthday:** The birthday of the person
- **Location_during_crime:** The location of the person during the crime
- **Eye_Color:** The eye color of the person
- **Skin_Color:** The skin color of the person
- **Hair_Color:** The hair color of the person
- **Height:** The height of the person in cm

**Witnesses:**

This dataset contains a statement from everyone in profiles, regarding suspicious things they saw during the time of the crime. Information that comes in this dataset are as follows:
- **ID:** A **Unique** identification for each person
- **Statement:** A statement from the person regarding a suspicious activity during the time of the crime

**Zichao's magic show:**

This dataset contains everyone that purchased a ticket for the show. **NOTE** that not everyone in profiles has purchased a ticket for this show. The Information contained in this dataset are as follows:
- **ID:** A **Unique** indentification for each person
- **Ticket_Number:** A **Unique** number that serves as an identification for the ticket bought

**Interrogation Statements:**

This datasets contains a statement from each person after they have been interrogated. **NOTE** assume that each statement is 100% true and that there are people who aren't in "Profiles" in this dataset. The Information contained in this dataset are as follows:

- **Name:** The name of the person being interrogated
- **Statement:** A statement the person gives after being interrogated

---

## Instructions for the Students:

In this assignment, you are tasked with solving the mystery of Dr. Fardina's stolen markers and finding the culprit.

* To accomplish this, you will be provided with **4 data sets** containing key information related to the incident. The list of the data sets name given below:

 - Profiles
 - Witnesses
 - Zichao's magic show
 - Interrogation Statements



* Your goal is to analyze these data sets to uncover the clues and evidences that will lead you to the thief.

**Be careful in following instructions as it may lead to you losing points :(**

Take a moment to look at summary of all the data sets below:

---

**Profiles:**

This dataset contains information regarding all the potential people that could have committed the crime.

**Witnesses:**

This dataset contains a statement from everyone in profiles, regarding suspicious things they saw during the time of the crime.

**Zichao's magic show:**

This dataset contains everyone that purchased a ticket for the show. **NOTE** that not everyone in profiles has purchased a ticket for this show. In this dataset, there is a column named  **"Ticket_Number"** which is an **Unique** number that serves as an identification for the ticket bought.

**Interrogation Statements:**

This datasets contains a statement from each person after they have been interrogated. **NOTE** assume that each statement is 100% true and that there are people who aren't in "Profiles" in this dataset.

---

## Q1) Lets begin by observing the profiles dataset


**TASK 1.1 (1 Point):**  In the code box below, load in "Profiles" dataset into a variable called **profiles_df** and display it





In [2]:
# Write your code here
profiles_df = pd.read_csv(r"C:\Users\asuto\Desktop\Master's\Padhai\DATA602 - Data Science\DATA602_HW\DATA602_FardinaAlam\HW\HW3\Profiles.csv")

**TASK 1.2 (1 Point):**  Next, you need to display the number of rows and columns in the "Profiles" dataset.

In [3]:
# Write your code here
profiles_df.shape

(348, 8)

**TASK 1.3 (1 Point):**  Next, you need to display the name of the columns in the "Profiles" dataset.

In [4]:
# Write your code here
profiles_df.columns

Index(['ID', 'Name', 'Birthday', 'Location_during_crime', 'Eye_Color',
       'Skin_Color', 'Hair_Color', 'Height'],
      dtype='object')

**TASK 1.4 (1 Point):**  Now, you need to display the types of the columns in the "Profiles" dataset.

In [5]:
# Write your code here
profiles_df.dtypes

ID                       object
Name                     object
Birthday                 object
Location_during_crime    object
Eye_Color                object
Skin_Color               object
Hair_Color               object
Height                    int64
dtype: object

**TASK 1.5 (2 Points):** Utilizing the .count() function on this dataset and set it to the variable **info** and then display it. Write any **potential problem** that you see with the dataset when observing the output of the count function:




In [6]:
## Write your code here
info = profiles_df.count()
info

ID                       348
Name                     348
Birthday                 348
Location_during_crime    280
Eye_Color                280
Skin_Color               348
Hair_Color               348
Height                   348
dtype: int64

In [7]:
348-280

68

*Write your observations here:*

There are several rows in the "Location_during_crime" and "Eye_Color" column which are null. 68 rows in particular.


**TASK 1.6 (4 Points):** Look back at the "Profiles" dataset. Using other pandas functions than count(), can you spot another potential problem with the dataset?

In [8]:
# Write your code here
profiles_df['Name'].nunique()

299

*Write your observation here:*

There are names which are duplicated(348 rows and 299 unique names). Indicating, that a few people may have had duplicate records in the profiles dataset.

## Q2) Cleaning the data set

**TASK 2.1 (10 Points):** Lets begin by fixing the Birthday column. You may notice that the dates are in inconsistent format!

- Pandas internally represent dates in 'datetime64' format which is a form of ISO 8601. Basically, dates are represented as yyyy-mm-dd.


- Fix all the dates in the Birthday Column.
- Display the "Profiles" dataframe after you are done.
- Note that the only 2 formats the dates are in is either yyyy/dd/mm or mm/dd/yyyy, you want to represent them as yyyy-mm-dd


**HINT:** some datetime functions might make things easier.
- https://www.geeksforgeeks.org/python-pandas-to_datetime/
- https://www.programiz.com/python-programming/datetime/strftime


In [9]:
%%time
def dateFnx(row):
    if len(row.split('/')[0]) == 4:
        return pd.to_datetime(row,format='%Y/%d/%m').strftime('%Y-%m-%d')
    return pd.to_datetime(row,format='%m/%d/%Y').strftime('%Y-%m-%d')

profiles_df.Birthday = pd.to_datetime(profiles_df.Birthday.apply(dateFnx))
profiles_df.head()

CPU times: total: 78.1 ms
Wall time: 60.5 ms


ID                   Name   Birthday Location_during_crime Eye_Color  \
0  P299            James Mason 1968-10-07          fourth floor       red   
1  P086  Christopher Rodriguez 1978-10-03          second floor    indigo   
2  P022           Misty Ashley 1989-08-10          fourth floor      blue   
3  P003           Amanda Brown 1995-10-27                 lobby     green   
4  P036          William Jones 1990-02-26          fourth floor      blue   

  Skin_Color Hair_Color  Height  
0     indigo     orange     191  
1     indigo      green     198  
2        red     violet     154  
3       blue     violet     158  
4     yellow     orange     184

**TASK 2.2 (5 Points):** Now that we have cleaned up the Birthday column, lets find the age of each person. Create a new column called "Age" that displays the age of the person **on the day of the crime.** Once you are done, display the
"Profiles" dataframe

In [10]:
# Write your code here
profiles_df['Age'] = (pd.to_datetime('2024-09-10',format='%Y-%m-%d') - profiles_df.Birthday).dt.days/365
profiles_df.head()

ID                   Name   Birthday Location_during_crime Eye_Color  \
0  P299            James Mason 1968-10-07          fourth floor       red   
1  P086  Christopher Rodriguez 1978-10-03          second floor    indigo   
2  P022           Misty Ashley 1989-08-10          fourth floor      blue   
3  P003           Amanda Brown 1995-10-27                 lobby     green   
4  P036          William Jones 1990-02-26          fourth floor      blue   

  Skin_Color Hair_Color  Height        Age  
0     indigo     orange     191  55.964384  
1     indigo      green     198  45.969863  
2        red     violet     154  35.109589  
3       blue     violet     158  28.893151  
4     yellow     orange     184  34.561644

**TASK 2.3 (10 Points):** Now that we have created the "age"; lets clean the 'Eye_color's column.

***Please read the directions carefully:***


- In order to fill in the missing eye_colors, we want to find the **most common eye_color** among those who have the **same hair and skin color**.
- For Instance, if we want to find the eye_color of someone who has yellow hair and violet skin:
 - we would first find everything who has the same hair and skin color, in this case it would be yellow hair and violet skin
 - and then we find the most common eye_color among them.
 - we then replace the missing eye_color with the eye color that is the most common among the people with the same hair and skin color.
- Watch out for redundant data!

- ***FIX: In the case that you find multiple modes, choose the color that appears first in alphabetical order***

 When you finish filling in the missing eye colors, please display the profiles_df


In [11]:
# Write your code here
eyeColorMap = profiles_df[['Skin_Color','Hair_Color','Eye_Color']].groupby(by=['Skin_Color','Hair_Color']).agg(pd.Series.mode)

def fnx(row):
    eye = eyeColorMap.loc[(row['Skin_Color'],row['Hair_Color'])].item()
    if type(eye) == type(np.array([])):
        return np.sort(eye)[0]
    else:
        return eye
idx = profiles_df[profiles_df.Eye_Color.isna()].index

profiles_df.loc[idx,'Eye_Color'] = profiles_df[profiles_df.Eye_Color.isna()].apply(fnx,axis=1)
profiles_df.head()

ID                   Name   Birthday Location_during_crime Eye_Color  \
0  P299            James Mason 1968-10-07          fourth floor       red   
1  P086  Christopher Rodriguez 1978-10-03          second floor    indigo   
2  P022           Misty Ashley 1989-08-10          fourth floor      blue   
3  P003           Amanda Brown 1995-10-27                 lobby     green   
4  P036          William Jones 1990-02-26          fourth floor      blue   

  Skin_Color Hair_Color  Height        Age  
0     indigo     orange     191  55.964384  
1     indigo      green     198  45.969863  
2        red     violet     154  35.109589  
3       blue     violet     158  28.893151  
4     yellow     orange     184  34.561644


## Q3) Finding the features of the culprit

Now that we have successfully cleaned all of our data, lets begin finding the features of the criminal. Take a moment to look at the **"Witnesses"** table. We don't necessarily want all the statements from every single witness.

**NOTE:** Take a look at the structure of each statement, notice how there are only a few variations of statements from the witnesses.

**TASK 3.1 (5 Points):** Find all witnesses that were near the crime scene
- First load in the Witness.csv file and set it to a variable called "**witness_df**"
- create a new dataframe called **"important_witnesses"** that contains the id and the witness statements of those who were near the scene of the crime
- display the new dataframe when you are finished
- **HINT:** look at the floor which the crime happened on.

In [12]:
# Write your code here
witness_df = pd.read_csv(r"C:\Users\asuto\Desktop\Master's\Padhai\DATA602 - Data Science\DATA602_HW\DATA602_FardinaAlam\HW\HW3\Witness.csv")
profiles_df['Location_during_crime'].value_counts()

Location_during_crime
fourth floor    79
third floor     73
second floor    71
lobby           57
Name: count, dtype: int64

In [13]:
fnx = lambda row:True if 'third' in row['Location_during_crime'] else False
tt = profiles_df[profiles_df['Location_during_crime'].notna()].apply(fnx,axis=1)
idx = tt[tt==True].index

important_witnesses = witness_df[witness_df.ID.isin(profiles_df.iloc[idx]['ID'])]
important_witnesses

ID                                          Statement
2    P184  I saw someone with indigo skin acting suspicious.
4    P001     I saw someone with red hair acting suspicious.
6    P283  I saw someone that looks to be 20 years old ac...
7    P190  I saw someone with orange eyes acting suspicious.
10   P071  I saw someone with violet eyes acting suspicious.
..    ...                                                ...
292  P275  I saw someone that looks to be 24 years old ac...
293  P148  I saw someone with violet eyes acting suspicious.
294  P053  I saw someone with orange skin acting suspicious.
295  P123  I saw someone that looks to be 20 years old ac...
297  P238  I saw someone with yellow hair acting suspicious.

[64 rows x 2 columns]

**TASK 3.2 (10 Points):** Find every potential eye color that the culprit could have

 - Put all the potential eye colors of the culprit in a list called **"potential_eye_colors"** and display it when you're done. potential_eye_colors should have **no duplicates.**

In [14]:
# Write your code here
potential_eye_colors = []

for _,row in important_witnesses.iterrows():
    statement = row['Statement']
    if 'eyes' in statement.split():
        flag = statement.split().index('eyes')
        color = statement.split()[flag-1]
        if color not in potential_eye_colors:
            potential_eye_colors.append(color)
potential_eye_colors

['orange', 'violet', 'red', 'blue']

**TASK 3.3 (5 Points):** Find every potential hair color that the culprit could have

 - Put all the potential hair colors of the culprit in a list called **"potential_hair_colors"** and display it when you're done

In [15]:
# Write your code here

potential_hair_colors = []

for _,row in important_witnesses.iterrows():
    statement = row['Statement']
    if 'hair' in statement.split():
        # print(statement)
        flag = statement.split().index('hair')
        color = statement.split()[flag-1]
        if color not in potential_hair_colors:
            potential_hair_colors.append(color)
potential_hair_colors

['red', 'violet', 'orange', 'yellow', 'green', 'indigo', 'blue']

**TASK 3.4 (5 Points):** Find every potential skin color that the culprit could have

 - Put all the potential skin colors of the culprit in a list called **"potential_skin_colors"** and display it when you're done

In [16]:
potential_skin_colors = []

for _,row in important_witnesses.iterrows():
    statement = row['Statement']
    if 'skin' in statement.split():
        # print(statement)
        flag = statement.split().index('skin')
        color = statement.split()[flag-1]
        if color not in potential_skin_colors:
            potential_skin_colors.append(color)
potential_skin_colors

['indigo', 'green', 'blue', 'red', 'violet', 'orange']

**TASK 3.5 (10 Points):** Find the potential height of the culprit

 - Put the smallest and largest potential heights of the culprit in the variables called **"smallest_height" & "largest_height"** and display them when you are done

In [17]:
# Write your code here
smallest_height,largest_height = profiles_df[(profiles_df.Eye_Color.isin(potential_eye_colors))|(profiles_df.Eye_Color.isin(potential_hair_colors))|(profiles_df.Eye_Color.isin(potential_skin_colors))].Height.agg(['min','max'])
smallest_height,largest_height

print("Smallest Height:",smallest_height)
print("Largest Height:",largest_height)

Smallest Height: 150
Largest Height: 200


**TASK 3.6 (8 Points):** Find the potential age of the culprit

 - Put the smallest and largest potential age of the culprit in the variables called **"smallest_age" & "largest_age"** and display them when you are done

In [18]:
# Write your code here
smallest_age,largest_age = profiles_df[(profiles_df.Eye_Color.isin(potential_eye_colors))|(profiles_df.Eye_Color.isin(potential_hair_colors))|(profiles_df.Eye_Color.isin(potential_skin_colors))].Age.agg(['min','max'])
smallest_age,largest_age

print("Smallest Height:",smallest_age)
print("Largest Height:",largest_age)

Smallest Height: 18.564383561643837
Largest Height: 66.34520547945205


## Q4) Finding who ones the ticket

**TASK 4.1 (2 Points):** Loading final dfs


Load in the **"Zichao's Magic Show.csv"** into a dataframe called **"show_df"** and the **"Interrogation_Statement.csv"** into a dataframe called **"interrogation_df"** respectively and display them when you are done

In [19]:
# Write your code here
show_df = pd.read_csv(r"C:\Users\asuto\Desktop\Master's\Padhai\DATA602 - Data Science\DATA602_HW\DATA602_FardinaAlam\HW\HW3\Zichao_s Magic Show.csv")
interrogation_df = pd.read_csv(r"C:\Users\asuto\Desktop\Master's\Padhai\DATA602 - Data Science\DATA602_HW\DATA602_FardinaAlam\HW\HW3\Interrogation_Statements.csv")

show_df.head()

ID  Ticket_Number
0  P132       50269975
1  P056       28120732
2  P250       88051289
3  P197       79545169
4  P012       83521845

In [20]:
interrogation_df.head()

Name                                          Statement
0  Jennifer Mccoy  Stealing? That's not in my nature at all. I re...
1     Karina Gray  Stealing? That's not in my nature at all. I re...
2  Craig Espinoza  I was busy with my project presentation in ano...
3  Michael Mendez  I had back-to-back exams that day. Stealing Dr...
4   Tara Thompson  I wasn't even on campus that day. Check the at...


**TASK 4.2 (8 Points):** Create a column called **Ticket_Color** for the **"show_df"** that displays all the colors of the ticket

Remember that there was a show ticket left at the crime scene, take a second to look through **"show_df"**. Try and find a way to identify how you can find the ticket color from the ticket number.

display the **"show_df"** when you are done

**HINT:** examine the **"interogation_df"**

In [21]:
# Write your code here
interrogation_df['Statement'].value_counts()

Statement
I was busy with my project presentation in another building. I didn't even step into Dr. Fardina's room.    120
Stealing? That's not in my nature at all. I respect everyone's property, especially Dr. Fardina's.          118
I had back-to-back exams that day. Stealing Dr. Fardina's marker would be the last thing on my mind!        115
Why would I need to steal a marker? I have plenty of my own. This accusation is absurd.                     106
I wasn't even on campus that day. Check the attendance logs; you won't find my name.                         38
We don't keep track of ticket colors, but my business partner Elizabeth Uiq might know how to find it!        1
Sum the individual #s in ticket numbers, look at the rightmost value, 0-2:B, 3-5:R , 6-7:G, 8-9:Y             1
Fardina's Markers were just too dang cool :( I saw them and I knew I had to have them. >:( arrest me          1
Name: count, dtype: int64

In [22]:
def fnx(row):
    summ = 0
    for i in str(row):
        summ += int(i)
    rightMost = summ%10
    if rightMost>=0 and rightMost<=2:
        return 'B'
    elif rightMost>=3 and rightMost<=5:
        return 'R'
    elif rightMost>=6 and rightMost<=7:
        return 'G'
    return 'Y'
    
show_df = show_df.assign(Ticket_Color = show_df['Ticket_Number'].apply(fnx))
show_df.head()

ID  Ticket_Number Ticket_Color
0  P132       50269975            R
1  P056       28120732            R
2  P250       88051289            B
3  P197       79545169            G
4  P012       83521845            G

## Q5) Finding the culprit

A helpful function might be the "iloc" function. Have a look at: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html

**TASK 5.1(10 Points):** Using all the data compiled from before find a person that matches all the features from the profile

- put the name of the culprit in a variable called **"culprit"**

In [23]:
# Write your code here
redTix = show_df[show_df.Ticket_Color=='R']
redTix.head()

ID  Ticket_Number Ticket_Color
0   P132       50269975            R
1   P056       28120732            R
5   P033       22058855            R
12  P111       87338906            R
15  P259       28602637            R

In [24]:
# potential_eye_colors,potential_hair_colors,potential_skin_colors,smallest_height,largest_height,smallest_age,largest_age

In [25]:
finDf = profiles_df[(profiles_df.Eye_Color.isin(potential_eye_colors)) & (profiles_df.Eye_Color.isin(potential_hair_colors)) & 
(profiles_df.Skin_Color.isin(potential_skin_colors)) & (profiles_df.Height>smallest_height) & (profiles_df.Height<largest_height) & 
(profiles_df.Age>smallest_age)& (profiles_df.Age<largest_age)& (profiles_df.ID.isin(redTix.ID))]

In [26]:
interrogation_df[interrogation_df.Name.isin(finDf.Name)]

Name                                          Statement
45         Ronald Boyer  Stealing? That's not in my nature at all. I re...
85            Steven Wu  I had back-to-back exams that day. Stealing Dr...
149  Cassandra Anderson  Why would I need to steal a marker? I have ple...
156     Patrick Farrell  I had back-to-back exams that day. Stealing Dr...
168        Pamela Salas  I had back-to-back exams that day. Stealing Dr...
184  Mr. Christian Boyd  I had back-to-back exams that day. Stealing Dr...
245    Rachel Alexander  Why would I need to steal a marker? I have ple...
247         Anne Butler  I was busy with my project presentation in ano...
269       Jason Hoffman  Why would I need to steal a marker? I have ple...
278       Jason Fischer  Why would I need to steal a marker? I have ple...
293       Colleen Hogan  I was busy with my project presentation in ano...
311       Heather Stone  I was busy with my project presentation in ano...
316      Abigail Nelson  Why would I need to steal a marker? I have ple...
322      Amanda Morales  I was busy with my project presentation in ano...
344         Malik Perez  I had back-to-back exams that day. Stealing Dr...
363       David Collins  I was busy with my project presentation in ano...
409         Michael Lee  I had back-to-back exams that day. Stealing Dr...
429      Paul Mejia DDS  I was busy with my project presentation in ano...
443       Charles Jones  I had back-to-back exams that day. Stealing Dr...
453       Stephanie Lam  I was busy with my project presentation in ano...
468           Ishan Kar  Fardina's Markers were just too dang cool :( I...
496        Adrian Black  I was busy with my project presentation in ano...

In [27]:
culprit = "Ishan Kar"

**TASK 5.2 (2 Points):** Check if your culprit is correct!

interrogate your culprit to see if you got the correct person

In [28]:
# Write your code here
interrogation_df[interrogation_df.Name==culprit]['Statement'].item()

"Fardina's Markers were just too dang cool :( I saw them and I knew I had to have them. >:( arrest me"

In [29]:
# Ishan Kar was the culprit. He stole the markers.